datafusion/src/physical_plan/metrics.rs

```diff
@@ -0,0 +1,461 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Metrics for recording information about execution
+
+pub mod wrappers;
+
+use std::{
+    fmt::{Debug, Display},
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+};
+
+use hashbrown::HashMap;
+
+use self::wrappers::{Count, Time};
+
+/// Structure for constructing metrics, counters, timers, etc
+pub struct MetricBuilder<'a> {
+    /// Location that the metric created by this builder will be added do
+    metrics: &'a SharedMetricsSet,
+
+    /// optional partition number
+    partition: Option<usize>,
+
+    /// arbitrary name=value pairs identifiying this metric
+    labels: Vec<Label>,
+}
+
+impl<'a> MetricBuilder<'a> {
+    /// Create a new `MetricBuilder` that will register the result of `build()` with the `metrics`
+    pub fn new(metrics: &'a SharedMetricsSet) -> Self {
+        Self {
+            metrics,
+            partition: None,
+            labels: vec![],
+        }
+    }
+
+    /// Add a label to the metric being constructed
+    pub fn with_label(mut self, label: Label) -> Self {
+        self.labels.push(label);
+        self
+    }
+
+    /// Add a label to the metric being constructed
+    pub fn with_new_label(
+        self,
+        name: impl Into<Arc<str>>,
```

Comment:
The use of `Arc<str>` for this, and labels, etc... is a little bit surprising to me. I'd expect these to in most cases be static strings, except in a couple of places where they might be unique dynamic strings (e.g. the output of format!). Perhaps `Cow<'static, str>` might be a better fit and would avoid allocations for static strings?

Reply:
Changed to `Cow<'static, str>` as suggested